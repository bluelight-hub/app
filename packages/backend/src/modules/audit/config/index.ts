export * from './audit-interceptor.config';
