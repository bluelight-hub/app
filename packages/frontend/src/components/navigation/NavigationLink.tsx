import React from 'react';
import { Link, LinkProps } from 'react-router-dom';
import { Button, Tooltip } from 'antd';
import { useActiveNavigation } from '@/hooks/useActiveNavigation';
import { useAuthorization } from '@/hooks/useAuthorization';

export interface NavigationLinkProps extends Omit<LinkProps, 'to'> {
  /** Ziel-URL */
  to: string;
  /** Anzuzeigender Text */
  children: React.ReactNode;
  /** Icon für den Link */
  icon?: React.ReactNode;
  /** Zusätzliche CSS-Klassen für aktive Links */
  activeClassName?: string;
  /** Soll exakte Übereinstimmung geprüft werden? */
  exact?: boolean;
  /** Ist der Link deaktiviert? */
  disabled?: boolean;
  /** Tooltip-Text */
  tooltip?: string;
  /** Button-Stil verwenden */
  asButton?: boolean;
  /** Button-Typ (falls asButton = true) */
  buttonType?: 'default' | 'primary' | 'ghost' | 'dashed' | 'link' | 'text';
  /** Button-Größe (falls asButton = true) */
  buttonSize?: 'small' | 'middle' | 'large';
  /** Callback für Klick-Events */
  onClick?: (event: React.MouseEvent<HTMLElement>) => void;
  /** Prüfe Zugriffsberechtigung automatisch */
  checkAccess?: boolean;
  /** Fallback-Komponente bei fehlender Berechtigung */
  unauthorizedFallback?: React.ReactNode;
}

/**
 * Erweiterte NavigationLink-Komponente mit automatischer Active-State-Erkennung
 * und Zugriffsberechtigungsprüfung
 */
const NavigationLink: React.FC<NavigationLinkProps> = ({
  to,
  children,
  icon,
  className = '',
  activeClassName = 'active',
  exact = false,
  disabled = false,
  tooltip,
  asButton = false,
  buttonType = 'default',
  buttonSize = 'middle',
  onClick,
  checkAccess = true,
  unauthorizedFallback = null,
  ...linkProps
}) => {
  const { isExactActive, isActiveOrChild, getLinkClasses } = useActiveNavigation();
  const { canAccessPath } = useAuthorization();

  // Prüfe Aktivitätsstatus
  const isActive = exact ? isExactActive(to) : isActiveOrChild(to);

  // Prüfe Zugriffsberechtigung
  const hasAccess = !checkAccess || canAccessPath(to);

  // Handle Click
  const handleClick = (event: React.MouseEvent<HTMLElement>) => {
    if (disabled || !hasAccess) {
      event.preventDefault();
      return;
    }

    if (onClick) {
      onClick(event);
    }
  };

  // Zeige Fallback bei fehlender Berechtigung
  if (!hasAccess && unauthorizedFallback) {
    return <>{unauthorizedFallback}</>;
  }

  // Generiere CSS-Klassen
  const linkClasses = getLinkClasses(to, `navigation-link ${className}`, `navigation-link--active ${activeClassName}`);

  // Zusätzliche Klassen für disabled/unauthorized
  const finalClasses = `
    ${linkClasses}
    ${disabled ? 'navigation-link--disabled' : ''}
    ${!hasAccess ? 'navigation-link--unauthorized' : ''}
    ${isActive ? 'navigation-link--current' : ''}
  `.trim();

  // Button-Stil
  if (asButton) {
    const buttonElement = (
      <Button
        type={isActive ? 'primary' : buttonType}
        size={buttonSize}
        icon={icon}
        disabled={disabled || !hasAccess}
        className={finalClasses}
        onClick={handleClick}
      >
        {children}
      </Button>
    );

    // Wrapper um Button für Navigation
    const buttonWithNavigation = (
      <Link
        to={to}
        {...linkProps}
        className="navigation-link-wrapper"
        onClick={handleClick}
        style={{ textDecoration: 'none' }}
      >
        {buttonElement}
      </Link>
    );

    return tooltip ? (
      <Tooltip title={tooltip} placement="top">
        {buttonWithNavigation}
      </Tooltip>
    ) : (
      buttonWithNavigation
    );
  }

  // Standard Link-Stil
  const linkElement = (
    <Link
      to={to}
      {...linkProps}
      className={finalClasses}
      onClick={handleClick}
      aria-current={isActive ? 'page' : undefined}
      aria-disabled={disabled || !hasAccess}
    >
      {icon && <span className="navigation-link__icon">{icon}</span>}
      <span className="navigation-link__text">{children}</span>
    </Link>
  );

  return tooltip ? (
    <Tooltip title={tooltip} placement="right">
      {linkElement}
    </Tooltip>
  ) : (
    linkElement
  );
};

export default NavigationLink;
