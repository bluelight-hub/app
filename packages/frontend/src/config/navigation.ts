import { IconType } from 'react-icons';
import {
    PiAmbulance,
    PiBell,
    PiBookOpen,
    PiChartPie,
    PiChecks,
    PiCheckSquareBold,
    PiClipboardBold,
    PiClipboardText,
    PiClock,
    PiCloud,
    PiDrone,
    PiFirstAid,
    PiGauge,
    PiGaugeBold,
    PiInfo,
    PiList,
    PiMapPinLineBold,
    PiMapTrifold,
    PiNavigationArrowBold,
    PiNotebook,
    PiNotePencilBold,
    PiPersonSimpleCircle,
    PiPlus,
    PiSquaresFour,
    PiSunDimBold,
    PiUserFocus,
    PiUsers,
    PiVideoCameraBold,
    PiWarningDiamond,
    PiWrench
} from 'react-icons/pi';

export type NavigationType = 'item' | 'group' | 'submenu';

export interface BaseNavigationItem {
    type: NavigationType;
    key: string;
    label: string;
    icon?: IconType;
}

export interface NavigationItemWithClick extends BaseNavigationItem {
    type: 'item';
    path: string;
}

export interface NavigationGroup extends BaseNavigationItem {
    type: 'group';
    children: (NavigationItemWithClick | NavigationSubmenu)[];
}

export interface NavigationSubmenu extends BaseNavigationItem {
    type: 'submenu';
    children: NavigationItemWithClick[];
}

export type NavigationItem = NavigationItemWithClick | NavigationGroup | NavigationSubmenu;

export const mainNavigation: NavigationItem[] = [
    {
        type: 'item',
        key: '/app',
        path: '/app',
        label: 'Dashboard',
        icon: PiGauge,
    },
    {
        type: 'group',
        key: 'einsatz',
        label: 'Einsatz',
        children: [
            {
                type: 'item',
                key: '/app/einsatztagebuch',
                path: '/app/einsatztagebuch',
                label: 'Einsatztagebuch',
                icon: PiNotebook,
            },
            {
                type: 'item',
                key: '/app/checklisten',
                path: '/app/checklisten',
                label: 'Checklisten',
                icon: PiCheckSquareBold,
            },
            {
                type: 'item',
                key: '/app/reminders',
                path: '/app/reminders',
                label: 'Wecker & Erinnerungen',
                icon: PiBell,
            },
            {
                type: 'submenu',
                key: 'kräfte',
                label: 'Kräfte',
                icon: PiPersonSimpleCircle,
                children: [
                    {
                        type: 'item',
                        key: '/app/kräfte',
                        path: '/app/kräfte',
                        label: 'Kräfteübersicht',
                        icon: PiSquaresFour,
                    },
                    {
                        type: 'item',
                        key: '/app/fahrzeuge',
                        path: '/app/fahrzeuge',
                        label: 'Fahrzeuge',
                        icon: PiAmbulance,
                    },
                    {
                        type: 'item',
                        key: '/app/einsatzkräfte',
                        path: '/app/einsatzkräfte',
                        label: 'Einsatzkräfte',
                        icon: PiUserFocus,
                    },
                    {
                        type: 'item',
                        key: '/app/rollen',
                        path: '/app/rollen',
                        label: 'Rollen',
                        icon: PiUsers,
                    },
                ],
            },
            {
                type: 'submenu',
                key: 'betroffene',
                label: 'Betroffene & Patienten',
                icon: PiFirstAid,
                children: [
                    {
                        type: 'item',
                        key: '/app/betroffene',
                        path: '/app/betroffene',
                        label: 'Betroffenenübersicht',
                        icon: PiSquaresFour,
                    },
                    {
                        type: 'item',
                        key: '/app/betroffene/aufnehmen',
                        path: '/app/betroffene/aufnehmen',
                        label: 'Aufnehmen',
                        icon: PiPlus,
                    },
                    {
                        type: 'item',
                        key: '/app/betroffene/verwalten',
                        path: '/app/betroffene/verwalten',
                        label: 'Verwalten',
                        icon: PiUsers,
                    },
                    {
                        type: 'item',
                        key: '/app/betroffene/manv',
                        path: '/app/betroffene/manv',
                        label: 'MANV',
                        icon: PiAmbulance,
                    },
                ],
            },
            {
                type: 'item',
                key: '/app/anforderungen',
                label: 'Anforderungen',
                icon: PiClipboardText,
                path: '/app/anforderungen',
            },
        ],
    },
    {
        type: 'group',
        key: 'lageübersicht',
        label: 'Lageübersicht',
        children: [
            {
                type: 'submenu',
                key: 'lagekarte',
                label: 'Lagekarte',
                icon: PiMapTrifold,
                children: [
                    {
                        type: 'item',
                        key: '/app/lagekarte',
                        path: '/app/lagekarte',
                        label: 'Übersicht Lagekarte',
                        icon: PiSquaresFour,
                    },
                    {
                        type: 'item',
                        key: '/app/lagekarte/letzte-eintraege',
                        path: '/app/lagekarte/letzte-eintraege',
                        label: 'Letzte Einträge',
                        icon: PiClock,
                    },
                    {
                        type: 'item',
                        key: '/app/lagekarte/dwd-wetterkarte',
                        path: '/app/lagekarte/dwd-wetterkarte',
                        label: 'DWD Wetterkarte',
                        icon: PiCloud,
                    },
                ],
            },
            {
                type: 'submenu',
                key: 'uav',
                label: 'UAV',
                icon: PiDrone,
                children: [
                    {
                        type: 'item',
                        key: '/app/uav',
                        path: '/app/uav',
                        label: 'UAV Übersicht',
                        icon: PiSquaresFour,
                    },
                    {
                        type: 'item',
                        key: '/app/uav/flugaufträge',
                        path: '/app/uav/flugaufträge',
                        label: 'Flugaufträge',
                        icon: PiDrone,
                    },
                    {
                        type: 'item',
                        key: '/app/uav/flugzonen',
                        path: '/app/uav/flugzonen',
                        label: 'Flugzonen',
                        icon: PiMapPinLineBold,
                    },
                    {
                        type: 'item',
                        key: '/app/uav/routen',
                        path: '/app/uav/routen',
                        label: 'Flugrouten',
                        icon: PiNavigationArrowBold,
                    },
                    {
                        type: 'item',
                        key: '/app/uav/wetter',
                        path: '/app/uav/wetter',
                        label: 'Wetter',
                        icon: PiSunDimBold,
                    },
                    {
                        type: 'item',
                        key: '/app/uav/stream',
                        path: '/app/uav/stream',
                        label: 'Live-Stream',
                        icon: PiVideoCameraBold,
                    },
                    {
                        type: 'item',
                        key: '/app/uav/telemetrie',
                        path: '/app/uav/telemetrie',
                        label: 'Telemetrie',
                        icon: PiGaugeBold,
                    },
                    {
                        type: 'item',
                        key: '/app/uav/protokoll',
                        path: '/app/uav/protokoll',
                        label: 'Flugprotokoll',
                        icon: PiClipboardBold,
                    },
                    {
                        type: 'item',
                        key: '/app/uav/bericht',
                        path: '/app/uav/bericht',
                        label: 'Flugbericht',
                        icon: PiNotePencilBold,
                    },
                ],
            },
        ],
    },
    {
        type: 'group',
        key: 'kommunikation-funk',
        label: 'Kommunikation & Funk',
        children: [
            {
                type: 'item',
                key: '/app/kanalliste',
                path: '/app/kanalliste',
                label: 'Kanallisten',
                icon: PiList
            },
            {
                type: 'item',
                key: '/app/kommunikationsverzeichnis',
                path: '/app/kommunikationsverzeichnis',
                label: 'Kommunikationsverzeichnis',
                icon: PiBookOpen
            },
            {
                type: 'item',
                key: '/app/fms',
                path: '/app/fms',
                label: 'FMS',
                icon: PiGaugeBold
            }
        ]
    },
    {
        type: 'group',
        key: 'einsatzinformationen',
        label: 'Einsatzinformationen',
        children: [
            {
                type: 'item',
                key: '/app/einsatzdaten',
                path: '/app/einsatzdaten',
                label: 'Einsatzdaten',
                icon: PiInfo,
            },
            {
                type: 'item',
                key: '/app/einsatzabschnitte',
                path: '/app/einsatzabschnitte',
                label: 'Einsatzabschnitte',
                icon: PiChartPie,
            },
            {
                type: 'item',
                key: '/app/schaden',
                label: 'Schäden',
                icon: PiWrench,
                path: '/app/schaden',
            },
            {
                type: 'item',
                key: '/app/gefahren',
                path: '/app/gefahren',
                label: 'Gefahren',
                icon: PiWarningDiamond,
            },
        ],
    },
    {
        type: 'item',
        key: '/app/notizen',
        path: '/app/notizen',
        label: 'Notizen',
        icon: PiChecks,
    },
];

// TODO: [FEATURE] Workspaces - vielleicht unnötig?
export const workspaces: NavigationItem[] = [
    {
        type: 'group',
        key: 'workspaces',
        label: 'Workspaces',
        children: [
        ]
    },
];

/**
 * Findet den Titel für die aktuelle Route in der Navigation
 * @param pathname Der aktuelle Pfad aus dem Router
 * @returns Den gefundenen Titel oder 'Dashboard' als Fallback
 */
export const findRouteTitle = (pathname: string): string => {
    const findInItems = (items: NavigationItem[]): string => {
        for (const item of items) {
            if (item.key === pathname) {
                return item.label;
            }
            if ('children' in item && item.children) {
                const found = findInItems(item.children);
                if (found) return found;
            }
        }
        return '';
    };

    return findInItems(mainNavigation) || 'Dashboard';
};
