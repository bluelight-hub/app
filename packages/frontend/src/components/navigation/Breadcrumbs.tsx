import React, { useMemo } from 'react';
import { Link, useLocation } from 'react-router-dom';
import { Breadcrumb } from 'antd';
import { HomeOutlined } from '@ant-design/icons';
import { RouteUtils } from '@/config/routes';
import { useAuthorization } from '@/hooks/useAuthorization';

export interface BreadcrumbItem {
  /** Label für den Breadcrumb */
  label: string;
  /** Pfad für die Navigation */
  path: string;
  /** Ist der Breadcrumb klickbar? */
  clickable: boolean;
  /** Icon für den Breadcrumb */
  icon?: React.ReactNode;
  /** Ist dies der aktuelle/letzte Breadcrumb? */
  current?: boolean;
}

export interface BreadcrumbsProps {
  /** Zusätzliche CSS-Klassen */
  className?: string;
  /** Custom Separator zwischen Breadcrumbs */
  separator?: React.ReactNode;
  /** Zeige Home-Icon beim ersten Element */
  showHomeIcon?: boolean;
  /** Maximum Anzahl der angezeigten Breadcrumbs */
  maxItems?: number;
  /** Custom Breadcrumb-Items (überschreibt automatische Generierung) */
  customItems?: BreadcrumbItem[];
  /** Callback wenn ein Breadcrumb geklickt wird */
  onItemClick?: (item: BreadcrumbItem) => void;
}

/**
 * Dynamische Breadcrumb-Komponente
 * Generiert automatisch Breadcrumbs basierend auf der aktuellen Route und Route-Metadaten
 */
const Breadcrumbs: React.FC<BreadcrumbsProps> = ({
  className,
  separator = '/',
  showHomeIcon = true,
  maxItems = 5,
  customItems,
  onItemClick,
}) => {
  const location = useLocation();
  const { canAccessPath } = useAuthorization();

  // Generiere Breadcrumb-Items basierend auf der aktuellen Route
  const breadcrumbItems = useMemo((): BreadcrumbItem[] => {
    // Verwende custom Items falls vorhanden
    if (customItems) {
      return customItems;
    }

    // Generiere automatisch basierend auf der Route
    const items = RouteUtils.generateBreadcrumbs(location.pathname);

    return items.map((item, index) => ({
      ...item,
      current: index === items.length - 1,
      icon: index === 0 && showHomeIcon ? <HomeOutlined /> : undefined,
    }));
  }, [location.pathname, customItems, showHomeIcon]);

  // Begrenze die Anzahl der angezeigten Items
  const displayItems = useMemo(() => {
    if (breadcrumbItems.length <= maxItems) {
      return breadcrumbItems;
    }

    // Zeige immer das erste und die letzten Items
    const start = breadcrumbItems.slice(0, 1);
    const end = breadcrumbItems.slice(-(maxItems - 2));

    return [
      ...start,
      {
        label: '...',
        path: '',
        clickable: false,
        current: false,
      },
      ...end,
    ];
  }, [breadcrumbItems, maxItems]);

  // Rendere einzelnen Breadcrumb-Item
  const renderBreadcrumbItem = (item: BreadcrumbItem, _index: number) => {
    const handleClick = () => {
      if (onItemClick) {
        onItemClick(item);
      }
    };

    // Nicht klickbare Items oder Ellipsis
    if (!item.clickable || item.label === '...' || item.current) {
      return (
        <span
          className={`${item.current ? 'font-medium text-gray-900' : 'text-gray-500'}`}
          onClick={item.clickable ? handleClick : undefined}
        >
          {item.icon && <span className="mr-1">{item.icon}</span>}
          {item.label}
        </span>
      );
    }

    // Prüfe Zugriffsberechtigung für den Link
    const canAccess = canAccessPath(item.path);

    if (!canAccess) {
      return (
        <span className="text-gray-400 cursor-not-allowed">
          {item.icon && <span className="mr-1">{item.icon}</span>}
          {item.label}
        </span>
      );
    }

    // Klickbarer Link
    return (
      <Link
        to={item.path}
        className="text-blue-600 hover:text-blue-800 transition-colors duration-200"
        onClick={handleClick}
      >
        {item.icon && <span className="mr-1">{item.icon}</span>}
        {item.label}
      </Link>
    );
  };

  // Konvertiere zu Ant Design Breadcrumb Items
  const antdItems = displayItems.map((item, index) => ({
    title: renderBreadcrumbItem(item, index),
    key: item.path || `item-${index}`,
  }));

  // Zeige keine Breadcrumbs wenn keine Items vorhanden
  if (displayItems.length === 0) {
    return null;
  }

  return (
    <div className={`breadcrumbs-container ${className || ''}`}>
      <Breadcrumb separator={separator} items={antdItems} className="text-sm" />
    </div>
  );
};

export default Breadcrumbs;
