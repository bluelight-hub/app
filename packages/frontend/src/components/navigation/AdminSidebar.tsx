import React, { useMemo } from 'react';
import { Avatar, Dropdown, Layout, Menu, Typography } from 'antd';
import { LogoutOutlined, MenuFoldOutlined, MenuUnfoldOutlined, SettingOutlined, UserOutlined } from '@ant-design/icons';
import { useNavigate } from 'react-router-dom';
import { RouteDefinition, RouteUtils } from '@/config/routes';
import { useAuthorization } from '@/hooks/useAuthorization';
import { useActiveNavigation } from '@/hooks/useActiveNavigation';
import NavigationLink from './NavigationLink';
import { useAuth } from '@/hooks/useAuth';
import { MenuItem } from '@/utils/navigationConverter.ts';

const { Sider } = Layout;
const { Text } = Typography;

export interface AdminSidebarProps {
  /** Ist die Sidebar zusammengeklappt? */
  collapsed?: boolean;
  /** Callback für Collapse-Status-Änderung */
  onCollapse?: (collapsed: boolean) => void;
  /** Zusätzliche CSS-Klassen */
  className?: string;
  /** Breite der Sidebar */
  width?: number;
  /** Breite der zusammengeklappten Sidebar */
  collapsedWidth?: number;
}

/**
 * Admin Navigation Sidebar mit rollenbasierter Menüfilterung
 */
const AdminSidebar: React.FC<AdminSidebarProps> = ({
  collapsed = false,
  onCollapse,
  className,
  width = 256,
  collapsedWidth = 80,
}) => {
  const navigate = useNavigate();
  const { user, logout } = useAuth();
  const { authorizedRoutes } = useAuthorization();
  const { currentPath } = useActiveNavigation();

  // Filtere und gruppiere autorisierte Routen für die Navigation
  const navigationItems = useMemo(() => {
    const routes = authorizedRoutes.filter((route) => route.showInNavigation);
    const grouped = RouteUtils.getRoutesByCategory();

    return Object.entries(grouped)
      .map(([category, categoryRoutes]) => {
        const authorizedCategoryRoutes = categoryRoutes.filter((route) =>
          routes.some((authRoute) => authRoute.id === route.id),
        );

        if (authorizedCategoryRoutes.length === 0) return null;

        return {
          category,
          routes: authorizedCategoryRoutes,
        };
      })
      .filter(Boolean);
  }, [authorizedRoutes]);

  // Konvertiere Routen zu Ant Design Menu Items
  const menuItems = useMemo(() => {
    const items: MenuItem[] = [];

    navigationItems.forEach((group, groupIndex) => {
      if (!group) return;

      // Füge Kategorie-Header hinzu (nur wenn nicht collapsed)
      if (!collapsed && group.routes.length > 0) {
        items.push({
          type: 'group',
          label: getCategoryDisplayName(group.category),
          key: `group-${group.category}`,
        });
      }

      // Füge Routen hinzu
      group.routes.forEach((route: RouteDefinition) => {
        items.push({
          key: route.path,
          icon: getRouteIcon(route.icon),
          label: (
            <NavigationLink
              to={route.path}
              checkAccess={false} // Bereits in authorizedRoutes gefiltert
              className="w-full"
            >
              {route.title}
            </NavigationLink>
          ),
          title: route.description,
        });
      });

      // Füge Divider zwischen Gruppen hinzu (aber nicht nach der letzten)
      if (groupIndex < navigationItems.length - 1 && !collapsed) {
        items.push({
          type: 'divider',
          key: `divider-${groupIndex}`,
        });
      }
    });

    return items;
  }, [navigationItems, collapsed]);

  // User Dropdown Menu
  const userMenuItems = [
    {
      key: 'profile',
      icon: <UserOutlined />,
      label: 'Profil',
      onClick: () => navigate('/admin/profile'),
    },
    {
      key: 'settings',
      icon: <SettingOutlined />,
      label: 'Einstellungen',
      onClick: () => navigate('/admin/settings'),
    },
    {
      type: 'divider' as const,
    },
    {
      key: 'logout',
      icon: <LogoutOutlined />,
      label: 'Abmelden',
      onClick: logout,
    },
  ];

  // Bestimme aktuell ausgewählte Menu Items
  const selectedKeys = useMemo(() => {
    const route = RouteUtils.findByPath(currentPath);
    return route ? [route.path] : [];
  }, [currentPath]);

  return (
    <Sider
      width={width}
      collapsedWidth={collapsedWidth}
      collapsed={collapsed}
      onCollapse={onCollapse}
      className={`admin-sidebar ${className || ''}`}
      style={{
        overflow: 'auto',
        height: '100vh',
        position: 'fixed',
        left: 0,
        top: 0,
        bottom: 0,
        zIndex: 1000,
      }}
    >
      {/* Header */}
      <div className="admin-sidebar__header p-4 border-b border-gray-200">
        <div className="flex items-center justify-between">
          {!collapsed && (
            <div className="flex items-center space-x-2">
              <div className="w-8 h-8 bg-blue-600 rounded-lg flex items-center justify-center">
                <span className="text-white font-bold text-sm">BH</span>
              </div>
              <div>
                <Text className="font-semibold text-white">Bluelight Hub</Text>
                <br />
                <Text className="text-xs text-gray-300">Admin Panel</Text>
              </div>
            </div>
          )}

          <button onClick={() => onCollapse?.(!collapsed)} className="p-1 rounded hover:bg-gray-700 text-white">
            {collapsed ? <MenuUnfoldOutlined /> : <MenuFoldOutlined />}
          </button>
        </div>
      </div>

      {/* Navigation Menu */}
      <div className="admin-sidebar__navigation flex-1">
        <Menu
          mode="inline"
          theme="dark"
          selectedKeys={selectedKeys}
          items={menuItems}
          inlineCollapsed={collapsed}
          className="border-none"
        />
      </div>

      {/* User Profile Footer */}
      <div className="admin-sidebar__footer p-4 border-t border-gray-700">
        {collapsed ? (
          <Dropdown menu={{ items: userMenuItems }} placement="topRight" trigger={['click']}>
            <Avatar icon={<UserOutlined />} className="cursor-pointer w-full flex items-center justify-center" />
          </Dropdown>
        ) : (
          <Dropdown menu={{ items: userMenuItems }} placement="topRight" trigger={['click']}>
            <div className="flex items-center space-x-3 cursor-pointer hover:bg-gray-700 p-2 rounded">
              <Avatar icon={<UserOutlined />} />
              <div className="flex-1 min-w-0">
                <Text className="text-white font-medium block truncate">{user?.name || 'Admin User'}</Text>
                <Text className="text-gray-300 text-xs block truncate">{user?.role || 'Administrator'}</Text>
              </div>
            </div>
          </Dropdown>
        )}
      </div>
    </Sider>
  );
};

// Helper Funktionen
function getCategoryDisplayName(category: string): string {
  const categoryNames: Record<string, string> = {
    overview: 'Übersicht',
    management: 'Verwaltung',
    system: 'System',
    monitoring: 'Überwachung',
    tools: 'Tools',
    analytics: 'Analytics',
    configuration: 'Konfiguration',
    other: 'Weitere',
  };

  return categoryNames[category] || category;
}

function getRouteIcon(iconName?: string): React.ReactNode {
  // Diese Funktion würde die entsprechenden Icons basierend auf dem Namen zurückgeben
  // Implementierung abhängig von der verwendeten Icon-Bibliothek

  const iconMap: Record<string, React.ReactNode> = {
    House: <UserOutlined />, // Placeholder - echte Icons verwenden
    Users: <UserOutlined />,
    Buildings: <UserOutlined />,
    Gear: <SettingOutlined />,
    FileText: <UserOutlined />,
    Database: <UserOutlined />,
    Code: <UserOutlined />,
    ChartBar: <UserOutlined />,
    Sliders: <SettingOutlined />,
  };

  return iconName ? iconMap[iconName] || <UserOutlined /> : <UserOutlined />;
}

export default AdminSidebar;
