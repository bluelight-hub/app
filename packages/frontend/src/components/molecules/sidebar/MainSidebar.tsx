import React from 'react';
import { NavigationItem } from '../../../config/navigation';
import DesktopSidebar from './DesktopSidebar';
import MobileSidebar from './MobileSidebar';

interface MainSidebarProps {
  isOpen?: boolean;
  onClose?: () => void;
  isMobile?: boolean;
  navigation: NavigationItem[];
}

/**
 * Main sidebar component that renders either the mobile or desktop sidebar
 * based on the provided props
 */
const MainSidebar: React.FC<MainSidebarProps> = ({
  isOpen = false,
  onClose = () => {},
  isMobile = false,
  navigation,
}) => {
  // Always render both sidebars, but let CSS and conditional rendering
  // determine which one is visible
  return (
    <>
      {/* Mobile sidebar - only visible when isMobile is true and isOpen is true */}
      {isMobile && <MobileSidebar isOpen={isOpen} onClose={onClose} navigation={navigation} />}

      {/* Desktop sidebar - always rendered but only visible on larger screens */}
      <DesktopSidebar navigation={navigation} />
    </>
  );
};

export default MainSidebar;
