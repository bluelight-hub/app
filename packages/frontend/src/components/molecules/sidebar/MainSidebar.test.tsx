import { render, screen } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { NavigationItem } from '../../../config/navigation';
import MainSidebar from './MainSidebar';

// Mock matchMedia
beforeEach(() => {
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: vi.fn().mockImplementation(query => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: vi.fn(),
            removeListener: vi.fn(),
            addEventListener: vi.fn(),
            removeEventListener: vi.fn(),
            dispatchEvent: vi.fn(),
        })),
    });
});

// Mock child components
vi.mock('./MobileSidebar', () => ({
    __esModule: true,
    default: vi.fn(({ isOpen, navigation }) => (
        isOpen ? <div data-testid="mobile-sidebar">{navigation.length} items</div> : null
    )),
}));

vi.mock('./DesktopSidebar', () => ({
    __esModule: true,
    default: vi.fn(({ navigation }) => (
        <div data-testid="desktop-sidebar">{navigation.length} items</div>
    )),
}));

const mockNavigation: NavigationItem[] = [
    {
        type: 'item',
        key: '/test',
        path: '/test',
        label: 'Test Item',
    }
];

describe('MainSidebar', () => {
    // Unit Tests
    it('should render desktop sidebar by default', () => {
        render(
            <MainSidebar
                navigation={mockNavigation}
            />
        );

        // Desktop sidebar should always be rendered
        expect(screen.getByTestId('desktop-sidebar')).toBeInTheDocument();

        // Mobile sidebar should not be rendered by default
        expect(screen.queryByTestId('mobile-sidebar')).not.toBeInTheDocument();
    });

    it('should render mobile sidebar when isMobile is true and isOpen is true', () => {
        render(
            <MainSidebar
                isOpen={true}
                isMobile={true}
                navigation={mockNavigation}
            />
        );

        // Both sidebars should be rendered
        expect(screen.getByTestId('desktop-sidebar')).toBeInTheDocument();
        expect(screen.getByTestId('mobile-sidebar')).toBeInTheDocument();
    });

    it('should not render mobile sidebar when isMobile is true but isOpen is false', () => {
        render(
            <MainSidebar
                isOpen={false}
                isMobile={true}
                navigation={mockNavigation}
            />
        );

        // Desktop sidebar should be rendered, but not mobile
        expect(screen.getByTestId('desktop-sidebar')).toBeInTheDocument();
        expect(screen.queryByTestId('mobile-sidebar')).not.toBeInTheDocument();
    });

    it('should pass navigation to both sidebars', () => {
        render(
            <MainSidebar
                isOpen={true}
                isMobile={true}
                navigation={mockNavigation}
            />
        );

        // Verify navigation prop was passed correctly to both sidebars
        expect(screen.getByTestId('desktop-sidebar')).toHaveTextContent('1 items');
        expect(screen.getByTestId('mobile-sidebar')).toHaveTextContent('1 items');
    });

    it('should have correct defaults for optional props', () => {
        const { rerender } = render(
            <MainSidebar
                navigation={mockNavigation}
            />
        );

        // Default: Only desktop sidebar is rendered
        expect(screen.queryByTestId('mobile-sidebar')).not.toBeInTheDocument();

        // When only isMobile=true but isOpen remains default (false)
        rerender(
            <MainSidebar
                isMobile={true}
                navigation={mockNavigation}
            />
        );

        // Mobile sidebar still shouldn't be visible with default isOpen=false
        expect(screen.queryByTestId('mobile-sidebar')).not.toBeInTheDocument();
    });

    // Snapshot test
    it('should match snapshot', () => {
        const { container } = render(
            <MainSidebar
                navigation={mockNavigation}
            />
        );

        expect(container).toMatchSnapshot();
    });
}); 