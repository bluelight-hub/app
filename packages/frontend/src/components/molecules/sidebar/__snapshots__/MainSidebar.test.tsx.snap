// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MainSidebar > should match snapshot 1`] = `
<div>
  <div
    data-testid="desktop-sidebar"
  >
    1
     items
  </div>
</div>
`;
